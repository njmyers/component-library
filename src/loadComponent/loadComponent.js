// @flow
import * as React from 'react';

type Props = {
  loading: React.ComponentType<{}>,
  error: React.ComponentType<{}>,
};

type State = {
  status: string,
  component: React.ComponentType<{}>,
};

type Module = {
  default: React.ComponentType<{}>,
};

/**
 * Asynchronous loader for react components and webpack
 * @param {function} importComponent function that returns a module
 */
function loadComponent(importComponent: () => Promise<Module>, options: Props) {
  class AsyncComponent extends React.Component<Props, State> {
    static defaultProps = {
      error: () => 'error',
      loading: () => 'loading',
    };

    state = {
      status: 'loading',
      component: () => null,
    };

    async componentDidMount() {
      try {
        const { default: component } = await importComponent();

        this.setState({
          status: 'resolved',
          component,
        });
      } catch (e) {
        if (process.env.NODE_ENV === 'developement') console.error(e);
        this.setState({ status: 'error' });
      }
    }

    render() {
      const Wrapped: React.ComponentType<{}> = this.state.component;
      const ErrorComponent = this.props.error;
      const Loading = this.props.loading;

      switch (this.state.status) {
        case 'error':
          return <ErrorComponent />;
        default:
        case 'loading':
          return <Loading />;
        case 'resolved':
          return <Wrapped {...this.props} />;
      }
    }
  }

  // overwrite fallback components with options object
  AsyncComponent.defaultProps = {
    ...AsyncComponent.defaultProps,
    ...options,
  };

  return AsyncComponent;
}

export default loadComponent;
