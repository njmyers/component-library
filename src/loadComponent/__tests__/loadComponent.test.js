import React from 'react';
import TestComponent from '../TestComponent';
import loadComponent from '../loadComponent';

import renderer from 'react-test-renderer';

const timeout = (time = 500) =>
  new Promise((res, rej) => {
    setTimeout(() => res(), time);
  });

describe('tests for asynchronously loading a component', () => {
  test('it displays a loader while loading a component', () => {
    const Loading = () => 'loading';
    const LoadedTest = loadComponent(() => import('../TestComponent'));
    const loadedComponent = renderer.create(<LoadedTest />);
    const component = renderer.create(<Loading />);

    expect(loadedComponent.toJSON()).toBe(component.toJSON());
    expect(loadedComponent.toJSON()).toMatchSnapshot();
  });

  test("it displays an error when imported file doesn't exist", () => {
    const ErrorComponent = () => 'error';
    const LoadedTest = loadComponent(() => import('../NotExist'));
    const loadedComponent = renderer.create(<LoadedTest />);
    const component = renderer.create(<ErrorComponent />);

    return timeout(100).then(() => {
      expect(loadedComponent.toJSON()).toBe(component.toJSON());
      expect(loadedComponent.toJSON()).toMatchSnapshot();
    });
  });

  test('it loads a component', () => {
    const LoadedTest = loadComponent(() => import('../TestComponent'));
    const loadedComponent = renderer.create(<LoadedTest />);
    const component = renderer.create(<TestComponent />);

    return timeout(100).then(() => {
      expect(loadedComponent.toJSON()).toMatchObject(component.toJSON());
      expect(loadedComponent.toJSON()).toMatchSnapshot();
    });
  });

  test('it passes props', () => {
    const LoadedTest = loadComponent(() => import('../TestComponent'));
    const loadedComponent = renderer.create(<LoadedTest value="value" />);
    const component = renderer.create(<TestComponent value="value" />);

    return timeout(100).then(() => {
      expect(loadedComponent.toJSON()).toMatchObject(component.toJSON());
      expect(loadedComponent.toJSON()).toMatchSnapshot();
    });
  });

  test('it overwrites custom loader', () => {
    const CustomLoader = () => 'some loading thing';
    const LoadedTest = loadComponent(() => import('../TestComponent'), {
      loading: CustomLoader,
    });

    const loadedComponent = renderer.create(<LoadedTest />);
    const component = renderer.create(<CustomLoader />);

    expect(loadedComponent.toJSON()).toBe(component.toJSON());
    expect(loadedComponent.toJSON()).toMatchSnapshot();
  });
});
