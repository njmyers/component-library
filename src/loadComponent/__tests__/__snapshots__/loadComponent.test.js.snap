// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for asynchronously loading a component it displays a loader while loading a component 1`] = `"loading"`;

exports[`tests for asynchronously loading a component it displays an error when imported file doesn't exist 1`] = `"error"`;

exports[`tests for asynchronously loading a component it loads a component 1`] = `
<p>
  Test Component
</p>
`;

exports[`tests for asynchronously loading a component it overwrites custom loader 1`] = `"some loading thing"`;

exports[`tests for asynchronously loading a component it passes props 1`] = `
<p>
  value
</p>
`;
